clash-cores/src/Clash/Cores/UART.hs

```diff
@@ -0,0 +1,207 @@
+{-|
+  Copyright   :  (C) 2021, LUMI GUIDE FIETSDETECTIE B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  Rowan Goemans <goemansrowan@gmail.com>
+
+  UART transmitter and receiver module
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE NumericUnderscores #-}
+
+module Clash.Cores.UART(HzToPeriod, uartTx, uartRx, uart) where
+
+import Clash.Prelude
+
+-- | Convert frequency in hertz to period in picoseconds
+type HzToPeriod (rate :: Nat) = Div 1_000_000_000_000 rate
+
+-- | State used for the uart transmitter
+data UartTxState (period :: Nat)
+   = UartTxIdle
+   -- ^ Not transmitting anything
+   | UartTxTransmitting
+   -- ^ Transmitting a word
+       { uartTxWord :: BitVector 10 -- All databits that need to be transmitted
+                                    -- Includes start and stop bit
+       , uartTxPeriodCounter :: Index period -- Counter that tracks how many
+                                             -- cycles we need to hold the
+                                             -- current bit on the tx line
+       , uartTxWordCounter :: Index 10       -- Which bit we are transmitting
+       }
+  deriving (Show, Generic, NFDataX)
+
+-- | Create a word that includes start a stop bits from a raw word
+createTxWord
+  :: BitVector 8
+  -- ^ The word to transmit
+  -> BitVector 10
+  -- ^ The word including start and stop bit
+createTxWord d = (1 :: (BitVector 1)) ++# d ++# (0 :: (BitVector 1))
+
+-- | State transition function for the uart transmitter
+uartTxT
+  :: forall (period :: Nat)
+   . KnownNat period
+  => UartTxState period
+  -- ^ The current state
+  -> Maybe (BitVector 8)
+  -- ^ The word to transmit
+  -> ( UartTxState period -- The next state
+     , ( Bit -- The output bit to the tx line
+       , Bool -- An acknowledgement that the word to transmit has been received
+       )
+     )
```

Comment:
Could you move the documentation to a single line and add `-- ^`? That way it'll show up in the Haddock.

Reply:
This shows up a couple of times in this PR, so same for the others :)

Reply:
Fixed